### PR TITLE
Java bindings/interface support

### PR DESCRIPTION
**Pull request by realvictorprm**
Should one day fix https://github.com/digital-asset/daml/issues/11350

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/JavaBackend.scala [hunk 1]
```diff
@@ -52,13 +53,28 @@ private[codegen] object JavaBackend extends Backend with StrictLogging {
...
+//      case (res, module: ModuleWithContext) =>
+//        val templateNames = module.interface.astInterfaces.keys
+//          .map { name =>
+//            ClassName.bestGuess(inner.fullyQualifiedName(name, None, packagePrefixes))
+//          }
+//        res ++ templateNames
+//      case (res, _) => res
+//    })
+//  }
+
```
Comment (realvictorprm) [inline language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/JavaBackend.scala:70]: delete

File: language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/InterfaceClass.scala [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+            packagePrefixes,
+          )
+        )
+        .addType(generateContractClass(interfaceName))
+        .build()
+      logger.debug("End")
+      templateType
+    }
+
+  private val idFieldName = "id"
```
Comment (realvictorprm) [inline language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/InterfaceClass.scala:72]: Part of this stuff here badly needs some refactoring out into something else because this is mostly duplicating code from `TemplateClass.scala`. I need to figure out first though, which parts are easy to share between both classes, because the duplicated methods in here are slightly but importantly different.

File: language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/TemplateClass.scala [hunk 3]
```diff
@@ -333,6 +334,18 @@ private[inner] object TemplateClass extends StrictLogging {
...
+      val simpleName = interfaceName.qualifiedName.name.segments.last
+      idClassBuilder.addMethod(
+        MethodSpec
+          .methodBuilder(s"to$simpleName")
+          .addModifiers(Modifier.PUBLIC)
+          .addStatement(s"return new $name.ContractId(this.contractId)")
+          .returns(ClassName.bestGuess(s"$name.ContractId"))
+          .build()
+      )
+    }
```
Comment (realvictorprm) [inline language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/TemplateClass.scala:348]: The beautiful code that makes it possible to convert Childs to Parent :P

File: ledger-service/http-json/src/main/scala/com/digitalasset/http/PackageService.scala [hunk 4]
```diff
@@ -311,7 +311,7 @@ object PackageService {
       interface: iface.Interface
   ): Map[(TemplateId.RequiredPkg, Choice), iface.Type] =
     interface.typeDecls.flatMap {
-      case (qn, iface.InterfaceType.Template(_, iface.DefTemplate(choices, _, _))) =>
+      case (qn, iface.InterfaceType.Template(_, iface.DefTemplate(choices, _, _, _))) =>
```
Comment (akrmn) [inline ledger-service/http-json/src/main/scala/com/digitalasset/http/PackageService.scala:314]: I'm not sure how `getChoices` is used, but keep in mind that this implementation will not return any choices inherited from interfaces

Comment (akrmn) [reply ledger-service/http-json/src/main/scala/com/digitalasset/http/PackageService.scala:314]: ah, ignore me, I see now that resolved interface choices get appended to the main list by `resolveChoices`

File: language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/InterfaceClass.scala [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+import com.daml.ledger.javaapi
+import com.daml.lf.codegen.backend.java.JavaEscaper
+import com.daml.lf.data.Ref.{ChoiceName, PackageId, QualifiedName}
+import com.daml.lf.iface.{DefInterface, TemplateChoice}
+import com.squareup.javapoet._
+import com.typesafe.scalalogging.StrictLogging
+
+import javax.lang.model.element.Modifier
+
+object InterfaceClass extends StrictLogging {
```
Comment (stefanobaghino-da) [inline language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/InterfaceClass.scala:15]: If you have duplicate definitions between here and `TemplateClass` consider pulling them out and calling them.

File: language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/TemplateClass.scala [hunk 6]
```diff
@@ -86,10 +88,11 @@ private[inner] object TemplateClass extends StrictLogging {
     ParameterizedTypeName.get(ClassName.get(classOf[Optional[_]]), name)
   private def setOfStrings = ParameterizedTypeName.get(classOf[java.util.Set[_]], classOf[String])
 
-  private def generateContractClass(
+  private[inner] def generateContractClass(
       templateClassName: ClassName,
       key: Option[Type],
       packagePrefixes: Map[PackageId, String],
+      isForInterface: Boolean,
```
Comment (stefanobaghino-da) [inline language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/TemplateClass.scala:95]: I would probably change the processing to generate a builder, pass it around and decorate it with what matters for different cases rather than burying the logic deep into this and gating it with a flag.

File: language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/InterfaceClass.scala [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+      val packageSegments = module.segments.slowAppend(name.segments).toSeq.dropRight(1)
+      // consider the last name segment to be the java class name
+      val className = name.segments.toSeq.takeRight(1)
+
+      val packageName = packageSegments.map(_.toLowerCase)
+
+      (packageName ++ className)
+        .filter(_.nonEmpty)
+        .map(JavaEscaper.escapeString)
+        .mkString(".")
```
Comment (stefanobaghino-da) [inline language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/InterfaceClass.scala:30]: I see that you copied this from another place, but I believe there's a better way to do this. As a further note, this is possibly an instance of something that can be factored out in its own method (although I would stick to simple functions and not use inheritance).
```suggestion
      val packageName (module.segments.iterator ++ name.segments.iterator).init.map(_.toLowerCase)
      // consider the last name segment to be the java class name
      val className = Iterator.single(name.segments.last)

      (packageName ++ className)
        .filter(_.nonEmpty)
        .map(JavaEscaper.escapeString)
        .mkString(".")
```
`ImmArray.last` is `O(1)` (should be documented alongside the method) and using iterators we can avoid materializing the collection and copying stuff around unnecessarily. Please do double check that my code compiles, I wrote it from within Github.

**Review by stefanobaghino-da — DISMISSED** [review]
I'm surprised that the documentation needed no change, can you please double check that?

**Comment by realvictorprm** [comment]
> I'm surprised that the documentation needed no change, can you please double check that?

I forgot about adding stuff to the documentation, many thanks for the reminder!

File: language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/TemplateClass.scala (hunk 6 above)
Comment (realvictorprm) [reply language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/TemplateClass.scala:95]: This probably could be a viable solution here to reduce the code duplication between InterfaceClass & TemplateClass :thinking:

File: language-support/java/codegen/BUILD.bazel [hunk 8]
```diff
@@ -315,7 +315,9 @@ daml_compile(
     name = "ledger-tests-model",
     srcs = glob(["src/ledger-tests/daml/**/*.daml"]),
     enable_scenarios = True,
-    target = lf_version_configuration.get("default"),
+    # FIXME: https://github.com/digital-asset/daml/issues/12051
+    #  replace "dev" by "default", once interfaces are stable.
```
Comment (S11001001) [inline language-support/java/codegen/BUILD.bazel:319]:
```suggestion
    #  remove target argument once interfaces are stable. 
```

File: language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/CodeGenRunner.scala [hunk 9]
```diff
@@ -79,10 +79,19 @@ object CodeGenRunner extends StrictLogging {
...
+    val environmentInterface =
+      EnvironmentInterface.fromReaderInterfaces(interfaces.head, interfaces.tail: _*)
     val prefixes = interfacesAndPrefixes.collect { case (_, (key, Some(value))) =>
       (key, value)
     }.toMap
-    (interfaces, prefixes)
+    val fullyResolvedInterfaces =
+      interfaces.map(
+        _.resolveChoices(
+          Function unlift environmentInterface.astInterfaces.get,
```
Comment (S11001001) [inline language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/CodeGenRunner.scala:90]:
```suggestion
          environmentInterface.astInterfaces,
```

File: language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/IdClass.scala [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+        typeCon: TypeCon,
+        identifierToType: Map[QualifiedName, InterfaceType],
+        packageId: PackageId,
+    ): Option[Record.FWT] = {
+      // TODO: at the moment we don't support other packages Records because the codegen works on single packages
+      if (typeCon.name.identifier.packageId == packageId) {
+        identifierToType.get(typeCon.name.identifier.qualifiedName) match {
+          case Some(InterfaceType.Normal(DefDataType(_, record: Record.FWT))) =>
+            Some(record)
+          case _ => None
```
Comment (S11001001) [inline language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/IdClass.scala:95]:
```suggestion
        identifierToType.get(typeCon.name.identifier.qualifiedName) collect {
          case InterfaceType.Normal(DefDataType(_, record: Record.FWT)) =>
            record
```

**Review by S11001001 — APPROVED** [review]
Seconding @stefanobaghino-da's note about function factoring, otherwise, very nice.

File: language-support/java/codegen/BUILD.bazel (hunk 8 above)
Comment (S11001001) [reply language-support/java/codegen/BUILD.bazel:319]: Also added to #13296.

File: daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/Interface.scala [hunk 11]
```diff
@@ -79,17 +79,30 @@ final case class Interface(
     * }}}
     */
   def resolveChoices(
-      findInterface: PartialFunction[Ref.TypeConName, DefInterface.FWT]
+      findInterface: PartialFunction[Ref.TypeConName, DefInterface.FWT],
+      failIfUnresolvedChoicesLeft: Boolean = false,
```
Comment (ray-roestenburg-da) [inline daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/Interface.scala:83]: I would prefer an extra method, `resolveChoicesFailUnresolved` or another terrible name :-) over a boolean flag

File: language-support/codegen-common/src/test/scala/com/digitalasset/daml/codegen/UtilSpec.scala [hunk 12]
```diff
@@ -42,7 +42,7 @@ object UtilSpec {
 
   val trivialEnvInterfaceGen: Gen[EnvironmentInterface] = {
     val fooRec = Record(ImmArraySeq.empty)
-    val fooTmpl = InterfaceType.Template(fooRec, DefTemplate(Map.empty, Map.empty, None))
+    val fooTmpl = InterfaceType.Template(fooRec, DefTemplate(Map.empty, Map.empty, None, Seq.empty))
```
Comment (ray-roestenburg-da) [inline language-support/codegen-common/src/test/scala/com/digitalasset/daml/codegen/UtilSpec.scala:45]: Do you setup something somewhere where a template does implement interfaces? (just checking)

File: language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/JavaBackend.scala [hunk 13]
```diff
@@ -25,6 +25,7 @@ private[codegen] object JavaBackend extends Backend with StrictLogging {
     val tree = InterfaceTrees.fromInterfaces(interfaces)
     for ((decoderPkg, decoderClassName) <- conf.decoderPkgAndClass) {
       val templateNames = extractTemplateNames(tree, packagePrefixes)
+      // val interfaceNames = extractInterfaceNames(tree, packagePrefixes)
```
Comment (ray-roestenburg-da) [inline language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/JavaBackend.scala:28]: remove this line?

File: language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/InterfaceClass.scala [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+
+  def generate(
+      interfaceName: ClassName,
+      interface: DefInterface.FWT,
+      packagePrefixes: Map[PackageId, String],
+      packageId: PackageId,
+      interfaceId: QualifiedName,
+  ): TypeSpec =
+    TrackLineage.of("interface", interfaceName.simpleName()) {
+      logger.info("Start")
```
Comment (ray-roestenburg-da) [inline language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/InterfaceClass.scala:41]: Should this be info?

File: language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/ContractIdClass.scala [hunk 15]
```diff
@@ -0,0 +1,254 @@
...
+import com.daml.lf.iface.Type
+import com.squareup.javapoet._
+import scala.jdk.CollectionConverters._
+
+import java.util.Optional
+import javax.lang.model.element.Modifier
+
+object ContractIdClass {
+
+  def builder(
```
Comment (ray-roestenburg-da) [inline language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/ContractIdClass.scala:15]: Do you have tests for this builder?

**Review by ray-roestenburg-da — COMMENTED** [review]
Please add tests for the builder classes (I might have overlooked, let me know if they are already there).

**Comment by realvictorprm** [comment]
> Overall it looks good. (disclaimer: I still need to learn more about this code) Please add tests for the builder classes? (I might have overlooked, let me know if they are already there).

not exactly sure how you would see the builder to be tested. The overall testing code we have already ensures that the builders work correctly, because if they wouldn't, the tests wouldn't pass :smile:

**Comment by ray-roestenburg-da** [comment]
@victormueller-da I just noticed a lot of new code without tests. It might be easier to test specific edge cases and get more coverage by just testing the builders, providing input, asserting expected output. This can give better feedback on details, if anything regresses. If there is already enough coverage indirectly, than it is not needed.

**Comment by realvictorprm** [comment]
> @victormueller-da I just noticed a lot of new code without tests. It might be easier to test specific edge cases and get more coverage by just testing the builders, providing input, asserting expected output. This can give better feedback on details, if anything regresses. If there is already enough coverage indirectly, than it is not needed.

OK, thank you very much for letting me know!

**Comment by ray-roestenburg-da** [comment]
@victormueller-da I added some other minor comments, would be great if you could answer those.

File: language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/IdClass.scala (hunk 10 above)
Comment (realvictorprm) [reply language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/IdClass.scala]: thx

File: language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/InterfaceClass.scala (hunk 7 above)
Comment (realvictorprm) [reply language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/InterfaceClass.scala]: I refactored it such that the initial function this was coming from is being used instead of code duplication

File: language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/JavaBackend.scala (hunk 13 above)
Comment (realvictorprm) [reply language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/JavaBackend.scala]: done

File: language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/InterfaceClass.scala (hunk 14 above)
Comment (realvictorprm) [reply language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/InterfaceClass.scala:26]: honestly don't know, I focused on keeping the logic and doing only small refactoring (thus no semantic changes).

File: language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/ContractIdClass.scala (hunk 15 above)
Comment (realvictorprm) [reply language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/ContractIdClass.scala:30]: Funnily I can now say _yes_! I just renamed the `TemplateClassSpec` to `ContractIdClassBuilderSpec` (the names get horrible long xD), because the tests in there turned out to be only related to code which was moved out into the `ContractIdClass.scala`.

File: language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/IdClass.scala [hunk 16]
```diff
@@ -0,0 +1,197 @@
...
+  TypeCon,
+  TypeNumeric,
+  TypePrim,
+  TypeVar,
+}
+import com.squareup.javapoet._
+
+import javax.lang.model.element.Modifier
+
+object IdClass {
```
[the remark this answers is not in the document]
Comment (realvictorprm) [reply language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/IdClass.scala:24]: Not directly but indirectly covered by the integration tests.

File: language-support/codegen-common/src/test/scala/com/digitalasset/daml/codegen/UtilSpec.scala (hunk 12 above)
Comment (realvictorprm) [reply language-support/codegen-common/src/test/scala/com/digitalasset/daml/codegen/UtilSpec.scala]: The relevant tests are only in the integration tests, because the code needs to be both compile-able & semantically correct which I can only check via running it against a ledger.

File: daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/Interface.scala (hunk 11 above)
Comment (realvictorprm) [reply daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/Interface.scala:83]: I'm not entirely sure why we all seem to dislike single bool flags so much, however I can fulfill that request.

File: language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/ContractIdClass.scala (hunk 15 above)
Comment (ray-roestenburg-da) [reply language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/ContractIdClass.scala:30]: Great!

File: language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/InterfaceClass.scala (hunk 14 above)
Comment (ray-roestenburg-da) [reply language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/InterfaceClass.scala:26]: Ok, yeah it just looked like a testing left-over, changing a debug into info and then leaving it around.. (I'm not sure who reads these logs so maybe this could be changed to debug, maybe @stefanobaghino-da @stephencompall-DA know. (it's super minor of course)

File: language-support/codegen-common/src/test/scala/com/digitalasset/daml/codegen/UtilSpec.scala (hunk 12 above)
Comment (ray-roestenburg-da) [reply language-support/codegen-common/src/test/scala/com/digitalasset/daml/codegen/UtilSpec.scala]: Ok. Here the thought (during my review) was just that maybe you want to run a test where the DefTemplate implements interfaces. It's very minor.

File: daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/Interface.scala (hunk 11 above)
Comment (ray-roestenburg-da) [reply daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/Interface.scala:83]: It's a minor nitpick maybe but I do think it helps with the readability of the code, especially if someone would use it without naming the argument. (then you have to go look, what is this `true`)

**Comment by realvictorprm** [comment]
OK Documentation updated, can someone take a look over it and tell me how good/bad it is? :D

File: daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/Interface.scala (hunk 11 above)
Comment (ray-roestenburg-da) [reply daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/Interface.scala:83]: Can you make this change?

Comment (realvictorprm) [reply daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/Interface.scala:83]: Completely forgot about this one. I will make that change :)

Comment (realvictorprm) [reply daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/Interface.scala]: done :)

**Comment by realvictorprm** [comment]
If there are no further objections I'll merge it in a couple hours (leaving you all some time for going over it again if you want to).
